hudi-spark-datasource/hudi-spark/src/main/scala/org/apache/spark/sql/hudi/command/MergeIntoHoodieTableCommand.scala

```diff
@@ -348,7 +387,7 @@ case class MergeIntoHoodieTableCommand(mergeInto: MergeIntoTable) extends Hoodie
    * expressions to the ExpressionPayload#getInsertValue.
    */
   private def executeUpsert(sourceDF: DataFrame, parameters: Map[String, String]): Unit = {
-    val operation = if (StringUtils.isNullOrEmpty(parameters.getOrElse(PRECOMBINE_FIELD.key, ""))) {
+    val operation = if (StringUtils.isNullOrEmpty(parameters.getOrElse(PRECOMBINE_FIELD.key, "")) && updatingActions.isEmpty) {
```

Comment:
does this mean that, we can have a MIT command just w/ non matching conditions and we will use insert operation then? 
but this may not go thru the prepped code path right? 
only upsert will go through the prepped code path. 
It should not matter even if we go through insert code path(since everything is new record). just wanted to clarify.

Reply:
Previously we were just doing an insert operation if the precombine was not set. Because we now do the join and get the meta cols for the existing records, the resulting behavior was very strange.  If you look at the last test in TestMergeIntoTable2.scala "Test only insert for source table in dup key without preCombineField", what was happening is that we would end up with 5 records as the final result. The record id = 1 would match on both of the existing id 1 records in the table so the df that would get sent to spark sql writer would have 2 of that and 1 of id = 3 and then insert everything. The previous behavior was that only 1 additional copy would be inserted. This behavior does not make sense because first of all, there are 2 matching records, and secondly, it is "update set *" not "insert set *". I think this behavior makes more sense because we do an update now for matching records.

Reply:
I see. 
by this "The previous behavior was that only 1 additional copy would be inserted.", mean after 2nd MIT command as per master, we will send only 2 records with id 1 and 3 to HoodieSparkSqlWriter. can you confirm ?